src/Preg.php

```diff
@@ -22,6 +22,35 @@
  */
 final class Preg
 {
+    /**
+     * @param string|string[] $pattern
+     *
+     * @return string|string[]
+     */
+    public static function verify($pattern)
+    {
+        if (\is_array($pattern)) {
+            return \array_map(__METHOD__, $pattern);
+        }
+
```

Comment:
Can we have an early return if we already verified the pattern to avoid the too much overhead issue mentioned by @julienfalque ? Something like:
```
if (isset(self::$verifiedPatterns[$pattern])) {
    return;
}
```

Reply:
@kubawerlos We certainly are able to write that code. But I would like to mention that we should not.

I could now quote from Rob Pike 1989, however even more concretely from top of my head, the PCRE library already caches compiled patterns internally. At least this is what I remember from years ago when I wrote that CSS parser related lexer / parser thingy years ago you can still find on my Github page. It's PCRE regex based.,

/cc @julienfalque

Reply:
> But I would like to mention that we should not.

@ktomk and I still stand on my opinion that we should, please do quote from Rob Pike 1989 because I am eager to learn what is wrong with an early return, which we are using it this project a lot.

Reply:
@kubawerlos Don't get me wrong, I'm for early reaturns very often, if not always, for the structure of a method. The verification calls are at the top of each of those methods and for the error case they throw. This means the exit is pretty early.

Your suggestion as I understand it is to introduce an in-memory caching layer in form of an array for performance reasons which have not been demonstrated to the best of my current knowledge.

I told in a comment below that I have not measured the change. I know this is sort of weak from my end. This is purely because I don't know where the perf suite of the project is to do these measurements and I'm eager to learn more here.

As you have asked for the quote, it's mainly the rules 1+2 from the complexity section that sprung to my mind, even if not a C programmer I think it's worth to read the whole document which is not very large, but here is a limited quote:

> ...
> Complexity
>      Most programs are too complicated - that is, more complex than they need to be to solve their problems efficiently.  Why? Mostly it's because of bad design, but I will skip that issue here because it's a big one.  But programs are often complicated at the microscopic level, and that is something I can address here.
>      Rule 1.  You can't tell where a program is going to spend its time.  Bottlenecks occur in surprising places, so don't try to second guess and put in a speed hack until you've proven that's where the bottleneck is.
>     Rule 2.  Measure.  Don't tune for speed until you've measured, and even then don't unless one part of the code overwhelms the rest.
> ...

From: [Notes on Programming in C by Rob Pike; February 21, 1989](https://www.lysator.liu.se/c/pikestyle.html)

Even Pike write _rules_, please mind that this is in context of his writing. I do not want to say that these are rules in this project that I think should be followed, it is just that in my opinion this is generally good advice. And it's more generic.

In this case especially  - as I already wrote elsewhere - I think that PCRE caches the compiled patterns already.

So what I'm trying to say is, that if there is a performance issue, we need to find out first to address it. So if you have more data at hand, I really would love to here about it and if this is an issue which needs addressing I'm the last who says to not address it. But let's see if this is an issue first of all.

Reply:
This in-memory caching with single `if` would save a lot of calls for `preg_` functions - and the less calls to `preg_` function is worth to have it.

Reply:
@kubawerlos Even if so, which different would it make? Is this worth to consider at all?

Reply:
I don't know, it would be great to see some benchmarks with no `verify`, with cached version and with non-cached version.

Reply:
@kubawerlos Was afk for some days and then needed to be avail for others elsewhere first, so please excuse the feedback-less time. On topic, just let me know if you've found out. I don't know how things are measured in this project still.